tclpro/modules/compiler/doc/procomp.1
'\"
'\" Copyright (c) 1998-2000 by Ajuba Solutions
'\" All rights reserved.
'\"
'\" RCS: @(#) $Id: procomp.1,v 1.2 2000/08/08 02:06:33 welch Exp $
'\" 
.so man.macros
.TH procomp 1 "" TclPro "TclPro Applications"
.BS
'\" Note:  do not modify the .SH NAME line immediately below!
.SH NAME
procomp \- TclPro Compiler
.SH SYNOPSIS
\fBprocomp\fR ?\fIoptions\fR? \fIfile1\fR ?\fIfile2 ...\fR?
.SH "OPTIONS"
.PP
.IP "\fB\-help\fR" 15
Displays usage information, then exits without doing anything.
.IP "\fB\-nologo\fR" 15
Hides the copyright and version information. 
.IP "\fB\-force\fR" 15
Force overwrite; if the output file exists, delete
it before the run, to ensure that the compilation does not
fail because of permission errors. Because TclPro Compiler creates
the output file with the same permissions as the input file,
a \fB.tbc\fR file generated from a readonly \fB.tcl\fR file will also be
readonly; a second compilation will fail unless the \fB-force\fR flag
is given.
.IP "\fB\-out \fIname\fR" 15
Specifies that the output path is \fIname\fR.  If only one file is being
compiled, \fIname\fR may specify the complete output file name.  Otherwise
\fIname\fR must be the name of an existing directory to which all compiled
files will be written.
.IP "\fB\-prefix \fItype\fR" 15
Specifies the type of prefix string that should be prepended to the
output files.  The \fItype\fR value can be one of the following:
.RS
.TP
\fBnone\fR
Do not add a prefix string.
.TP
\fBauto\fR
Extract the prefix from the input file: everything
from the start of the file to the first non-comment
or empty line is prepended to the output file.
.TP
\fBtag\fR 
Extract the prefix from the input file: everything from the
start of the file to the first occurrence of a comment line starting
with the text \fBTclPro::Compiler::Include\fR is prepended to the
output file.
.PP
Any other value for \fItype\fR is assumed to be a path to a
file whose contents are prepended to the output file.
The default for this option is \fBauto\fR.
.RE
.IP "\fB\-quiet\fR" 15
Suppresses warnings about non-existent files.
.IP "\fB\-verbose\fR" 15
Verbose mode: messages are generated to log the progress of
the compilation.
.BE
.SH DESCRIPTION
.PP
The command \fBprocomp\fR is used to start TclPro Compiler.
For each input file, TclPro Compiler creates an output file containing
an equivalent script stored as bytecodes for the Tcl bytecode
interpreter in Tcl 8.0.3 or later.
.PP
By default, the resulting bytecode files will be created in the same
directory as the input files, but will be given a different extension
(\fB.tbc\fR).  The \fB\-out\fR flag can be used to specify an
alternate directory where the output files will be created.  If only
one file is being compiled, the \fB-out\fR flag can also be used to
specify the complete name of the output file, including the extension.
.SH "COMPILED FILES"
.PP
TclPro Compiler creates an internal representation of a compiled script
using the Tcl bytecode compiler that is built into the Tcl core,
performs some additional computations, and then emits a representation
of the compiled script to the output file. The
output file itself is a simple Tcl script that loads the bytecode
runtime package (the \fBtbcload\fR package), and then instructs the
runtime to load the internal state and execute the script.  Since the
compiled bytecodes are a valid Tcl script, you can use them anywhere
you could use the input script, such as the \fBsource\fR and
\fBeval\fR commands.  You can also invoke compiled scripts directly as
shell commands, just as you could the input script, as long as you
have created a proper header (see below).
.PP
One of the additional computations that TclPro Compiler performs is to
collect all procedure bodies that were defined in the script, and
create a compiled representation for them as well as the script (the
Tcl interpreter compiles procedure bodies lazily, whereas TclPro
Compiler forces a compilation early). Thus, scripts that contain
mostly procedure definitions are compiled into efficient
representations.
.PP
Because the compiled file is itself a Tcl script, there may be
situations where you will want to add some specialized setup code at
the start of the script. For example, in order to invoke a Tcl script
directly from a Unix shell, the first lines of the script should look
like the following (where \fBtclsh\fR may be replaced with the name of
a different interpreter, such as \fBwish\fR):
.PP
.CS
#!/bin/sh
# the next line restarts using tclsh \e
exec tclsh "$0" "$@"
.CE
.PP
By default, the compiler will
not generate such a construct. However, you can place such a header
in the Tcl source file (or in some other file) and instruct TclPro
Compiler to copy it to the output file using the \fB\-prefix\fR switch.
.SH "LIMITATIONS"
Currently the compiler will compile all code and code in any
control structure or procedure at the top level of a script.  However,
code, including procedures, inside of some commands (most notably
\fBnamespace eval\fR) will not be compiled.
.PP
The \fB.tbc\fR file can only be loaded in the \fBprotclsh\fR and
\fBprowish\fR interpreters or a standard Tcl 8.0.3 (or later) that has
the \fBtbcload\fR package available.  The \fBtbcload\fR package may be
distributed by TclPro users to their customers as needed.
.PP
Some constructs that are legal with regular Tcl scripts are not supported
with compiled scripts. For example, the \fBinfo body\fR command cannot
return the body of a compiled procedure, because that information is not
available; it will return instead a body that was automatically generated by
the \fBtbcload\fR package and that has no relation to the actual body of the
procedure. This new body contains a call to \fBerror\fR; thus, if we define
two procedures \fBA\fR and \fBB\fR like this:
.PP
.CS
proc A {args} { return "hello world $args" }
proc B {args} [info body A]
.CE
.PP
a call to \fBB TEST\fR from an uncompiled Tcl script will return the
string \fB"hello world TEST"\fR, as expected. A call from a compiled
script will cause an error to be generated instead.
.SH "EXAMPLES"
.PP
The simplest command line to compile a Tcl script is shown below:
.PP
.CS
procomp filename.tcl
.CE
.PP
Which will create a file \fBfilename.tbc\fR containing the bytecodes.
To compile all \fB.tcl\fR files in \fB/dir1\fR, type:
.PP
.CS
procomp /dir1/*.tcl
.CE
.PP
To place the output files in a different directory, use the
\fB\-out\fR flag to specify where they should be created:
.PP
.CS
procomp \-out /dir2 /dir1/*.tcl
.CE
.PP
To have TclPro Compiler extract a header from each \fB.tcl\fR file compiled
by looking for the \fBTclPro::Compiler::Include\fR tag,
and place it at the start of the corresponding \fB.tbc\fR file:
.PP
.CS
procomp \-out /dir2 \-prefix tag /dir1/*.tcl
.CE
.SH KEYWORDS
compiler bytecodes


tclpro/ChangeLog
2001-03-31  Andreas Kupries <andreas_kupries@users.sourceforge.net>

	* get-trfcrypt: Shortened in adaption to the new release of
	  Trfcrypt (2.0p3).

2001-03-31  Andreas Kupries <andreas_kupries@users.sourceforge.net>

	* get-setops: Shortened in adaption to the new release of setops
	  (1.0p2).

2001-03-28  Andreas Kupries <andreas_kupries@users.sourceforge.net>

	* get-trf:
	* get-memchan: Shortened in adaption to the new releases of trf
	  and memchan.

2001-03-16  Andreas Kupries <andreas_kupries@users.sourceforge.net>

	* get-trf:
	* get-expand:
	* get-trfcrypt:
	* get-setops:
	* get-memchan: Added code to move an existing archive before retrieving
	  the current one from the web. Without this 'wget' will create
	  names like 'foo.1', etc. instead of 'foo'. This causes the code
	  after the wget to actually use the old archive and not the new one.

2001-03-15  Karl Lehenbauer <karl@procplace.com>
	* snapshot/{README bundle.tcl listpackages.tcl postpro.tcl}:
	  Added snapshot directory and these new files.  

	  postpro.tcl grinds a SNAPSHOT file to produce a SNAPSHOT.merge 
	  file, containing concise information about what modules produced
	  what files in the output and defined what packages.

	  bundle.tcl takes a SNAPSHOT.merge file and produces numerous
	  zip files consisting of the files for each module, with
	  manual pages and demos split into separate zips.

	  listpackage.tcl if run by a tcl or wish shell will list all
	  of the packages available to that shell.

2001-03-14  Karl Lehenbauer <karl@procplace.com>

	* mkIndex.tcl.in: Altered Nativepath proc to work correctly with 
	  new Cygwin drive specification syntax when running on Windows.
	  (Was //d/foo -> D:/foo   Now also /cygdrive/d/foo -> D:/foo)

2001-03-07 Karl Lehenbauer <karl@procplace.com>
	* modules/install/src/setup.sh: Removed expectation that
	  uname resides in /bin -- it doesn't on FreeBSD and probably
	  others.

2001-03-06 Andreas Kupries <andreas_kupries@users.sourceforge.net>

	* configure.in:
	* module_data.tcl.in
	* modules/install/src/parts.in:
	* modules/install/src/cdparts.in: Added information to allow the
	  distribution of the tcl-based macro processor 'expand' by
	  William H. Duquette <will@wjduquette.com>. See
	  http://www.wjduquette.com/expand/index.html

	* get-expand: New file. Retrieves expand2.0 and patches the
	  distribution for usage by the huild framework.

2001-03-06 Andreas Kupries <andreas_kupries@users.sourceforge.net>

	* configure.in:
	* module_data.tcl.in
	* modules/install/src/parts.in:
	* modules/install/src/cdparts.in: Added information to allow the
	  compilation of 'memchan', 'trf' and 'trfcrypt'.

	* get-trf:
	* get-trfcrypt:
	* get-memchan: New files. Retrieve the current source
	  distributions of memchan, trf and trfcrypt, and patch them so
	  that the build framework can use them.

	* README.akupries: New file, short explanation of the above.

2001-02-23 Karl Lehenbauer  <karl@procplace.com>
	Documented and simplified soft status code in buildenv/buildModule.tcl.
	It is only needed for caching that configure has been done and
	whether or not "make install" has been done.  Now that's all
	we try to track.

	Added new make target, "make reinstall".  If you do this, it
	will build and install everything, ignoring any soft status
	information.

2001-02-20 Karl Lehenbauer  <karl@procplace.com>
	Modified dependency definitions in tclpro/module_data.tcl.in 
	and tclchecker/DEPENDENCIES.  In particular, tclchecker needs
	prowrap.  Noted in comments in module_data.tcl.in which modules'
	dependencies are overridden by explicit DEPENDENCIES files.

	Modified buildenv/buildModule.tcl and tclpro/Makefile.in to
	add the capability of skipping steps that have already been
	successfully performed by caching results in a SOFTSTATUS file
	managed by buildModule.tcl in the directory you run "make"
	from to build tclpro.  Thus if, for example, TclX has already
	been installed, it isn't installed over and over again whenever
	you do a make.  This is a little kludgey because it works by
	having, for example, the successful return of "make install" of
	TclX letting us know it worked, rather than looking for a specific
	file result.

	Also buildModule exits upon hitting an error, rather than proceeding.
	This should probably be made an option.

	Finally all TEA supported targets are now supported from the
	Makefile, so you can say stuff like "make tcl-install", 
	"make tcl_static-install", "make tcllib-clean", 
	"make tclhttpd-configure" etc.  Using make in this manner bypasses
	the soft status stuff described above.

2001-02-8  Brent Welch  <brent.welch@interwoven.com>
	TclProCheckout.tcl: Cleaned this up so it correctly detects if
	a read/write checkout is successful.  Added -no_action cmdline arg.

2001-02-6  Brent Welch  <brent.welch@interwoven.com>
	modules/install/src/genImage.tcl: Rooted out more references to license
	modules/install/src/text.tcl: Rooted out more references to license

2001-01-30  Brent Welch  <brent.welch@interwoven.com>
	README.FreeBSD: latest updates by Karl
	configure.in: Removed builds of documentation and mclistbox

2001-01-24  Brent Welch  <brent.welch@interwoven.com>
	(For Karl Lehenbauer)
	modules/wrapper/src/setup.tcl:
	Removed references to license code
	README.FreeBSD: added docs from Karl.

2001-01-24  Brent Welch  <brent.welch@interwoven.com>
	(For Karl Lehenbauer)
	modules/wrapper/src/startup.tcl:
	modules/debugger/src/startup.tcl: changed the way the
	    lcient calls are commented out so it shows up from a grep better
	modules/install/src/cdparts.in:
	modules/install/src/genImage.tcl:
	modules/install/src/gui.tcl:
	modules/install/src/install.tcl:
	modules/install/src/makeDist.tcl:
	    Fixes by Karl to get the installer images working on FreeBSD!
	modules/lmgr:
	    Removed the old, unused license manager.

2001-01-22  Brent Welch  <brent.welch@interwoven.com>
	* Makefile.in:
	Added a %-clean target that is equivalent to %-hose -

tclpro/modules/compiler/src/procomp.tcl
# procomp.tcl --
#
#  Front-end script for the TclPro bytecode compiler.
#  Compiles individual TCL scripts, or directories containing TCL scripts.
#  For individual scripts, the output file is either the one specified by the
#  -o flag, or it is created in the same location as the input file, but with
#  a different extension (.tbc). For directories, all files matching a given
#  pattern (which defaults to *.tcl) are compiled; the output files are either
#  placed in the directory specified by the -o flag, or in the same directory
#  as the input files. All output files have the same root as their input file,
#  and different extension.
#
#
# Copyright (c) 1998 by Scriptics Corporation.
#
# See the file "license.terms" for information on usage and redistribution
# of this file, and for a DISCLAIMER OF ALL WARRANTIES.
# 
# RCS: @(#) $Id: procomp.tcl,v 1.4 2001/02/08 21:40:55 welch Exp $

package require Tcl 8.0
package require compiler 1.0

package provide procomp 1.0

namespace eval procomp {
    namespace export run setLogProc setErrorProc

    # The Log and Error procedures are used to emit output and errors,
    # respectively.

    variable logProc defaultLogProc errorProc defaultErrorProc

    variable fileList {}
    variable headerType auto 
    variable byteCodeExtension .tbc
    variable forceWrite 0

    # This is the tag pattern that is recognized by the 'tag' header type.
    # The tag must appear on a comment line by itself.
    # The tagError variable controls the behaviour of the 'tag' lookup
    # routine; if it is 1, a missing tag trigger an error, if it is 0
    # a missing tag makes the proc return an empty header.

    variable tagValue TclPro::Compiler::Include
    variable tagPattern [format {^[ ]*#[ ]%s[ ]*$} $tagValue]
    variable tagError 0

    variable usage "Usage: [cmdline::getArgv0] ?options? path1 ?path2 ...?
  -force	force overwrite; if the output file exists, delete it.
  -help		print this help message.
  -nologo	suppress copyright banner.
  -out name	specifies that the output path is 'name'.  If only one file
		is being compiled, 'name' may specify the complete output file
		name.  Otherwise 'name' must be the name of an existing
		directory to which all compiled files will be written.
  -prefix type  specifies whether a prefix string should be prepended to the
		emitted output files, and if so how it should be generated
		'type' value can be one of the following:
		  none	do not add a prefix string
		  auto	extract the prefix from the input file: everything
			from the start of the file to the first non-comment
			or empty line is prepended to the output (default)
		  tag	extract the prefix from the input file: everything
			from the start of the file to the first occurrence
			of a comment line starting with the string 
			\"TclPro::Compiler::Include\" is prepended to output
		Any other value for type is assumed to be a path to a
		file that will be used as the output file prefix.
  -quiet	suppress warnings about non-existent files
  -verbose	verbose mode: messages are generated to log progress.
  pathN		one or more files to compile."
}

# procomp::run --
#
#  Runs the package.
#
# Arguments:
#  argList	(optional) the argument list. If not specified, the list is
#		initialized from ::argv.
#
# Results:
#  Returns 1 on success, 0 on failure.

proc procomp::run { {argList {}} } {
    variable fileList

    if {[init $argList] == 0} {
	return 0
    }

    foreach file $fileList {
	if {![fileCompile $file]} {
	    return 0
	}
    }
}

# procomp::setLogProc --
#
#  Sets the log procedure for the package.
#  The log procedure is called by the package when it emits output; it is
#  assumed to take a single argument, the string to log.
#
# Arguments:
#  procName	the name of the log procedure to use
#
# Results:
#  Returns the name of the log procedure that was in use before the call.

proc procomp::setLogProc { procName } {
    variable logProc
    set logProc $procName
}

# procomp::setErrorProc --
#
#  Sets the error log procedure for the package.
#  The error log procedure is called by the package when it emits an error
#  message; it is assumed to take a single argument, the string to log.
#
# Arguments:
#  procName	the name of the log procedure to use
#
# Results:
#  Returns the name of the log procedure that was in use before the call.

proc procomp::setErrorProc { procName } {
    variable errorProc
    set errorProc $procName
}

# procomp::init --
#
#  Initializes the package: checks the arguments, initializes the control
#  structures and checks them for consistency.
#
# Arguments:
#  argList	(optional) the argument list. If not specified, the list is
#		initialized from ::argv.
#
# Results:
#  Returns 1 on success, 0 on failure.

proc procomp::init { {argList {}} } {
    variable headerType
    variable byteCodeExtension [compiler::getBytecodeExtension]
    variable fileList
    variable headerValue
    catch {unset headerValue}
    variable outPath
    catch {unset outPath}
    variable tagValue
    variable logProc
    variable forceWrite
    variable usage

    set quiet 0

    if {[string compare $argList {}] == 0} {
	set argList $::argv
    }

    # if the user has entered nothing on the command line, display the
    # usage string and exit with error

    if {[llength $argList] < 1} {
	log $usage
	return 0
    }

    # initialize the control structures from the argument list: parse the
    # flags

    set optionList {
	? f force h help n nologo o.arg out.arg p.arg prefix.arg
	q quiet v verbose
    }

    set isVerbose 0
    set projectInfo::printCopyright 0
    while {[set err [cmdline::getopt argList $optionList opt arg]]} {
	if {$err < 0} {
	    log "[cmdline::getArgv0]: $arg (use \"-help\" for legal options)"
	    return 0
	} else {
	    switch -exact $opt {
		? -
		h -
		help {
		    log $usage
		    return 1
		}

		f -
		force {
		    set forceWrite 1
		}

		logo {
		    # This will turn on the copyright printing info
		    # when we check out license.

		    set projectInfo::printCopyright 1
		}

		o -
		out {
		    set outPath $arg
		}

		p -
		prefix {
		    set headerType $arg
		}

		q -
		quiet {
		    set quiet 1
		}

		v -
		verbose {
		    set isVerbose 1
		}
	    }
	}
    }

    # After processing args but before we go any futher we check
    # to see if the user has a valid license key.

    projectInfo::printCopyright "TclPro Compiler"

    if {!$isVerbose} {
	set logProc nullLogProc
    }

    # Ensure that the -out option is consistent with the list of files being
    # compiled.  Also ensure that we have at least one file.

    set fileList [cmdline::getfiles $argList $quiet]
    if {[llength $fileList] < 1} {
	logError "no files to compile"
	return 0
    } elseif {([llength $fileList] != 1) && [info exists outPath]} {
	if {![file isdir $outPath]} {
	    logError "-out must specify a directory when compiling more than one file"
	}
    }

    # now check the control structures for internal consistency.
    # First, the header type; if not one of the known types, it must be an
    # existing file.

    switch -- $headerType {
	none -
	auto -
	tag {
	}

	default {
	    if {[catch {open $headerType r} istm] == 1} {
		logError "error: bad header type: $istm"
		return 0
	    }

	    set headerValue [read $istm]
	    close $istm
	    append headerValue [format "# %s\n" $tagValue]
	}
    }

    return 1
}

# procomp::defaultLogProc --
#
#  Default logging procedure: writes $msg to stdout.
#
# Arguments:
#  msg		the message to log.
#
# Results:
#  None.

proc procomp::defaultLogProc { msg } {
    puts stdout $msg
}

# procomp::defaultErrorProc --
#
#  Default error logging procedure: writes $msg to stdout.
#
# Arguments:
#  msg		the error message to log.
#
# Results:
#  None.

proc procomp::defaultErrorProc { msg } {
    puts stdout $msg
}

# procomp::nullLogProc --
#
#  Null logging procedure: does nothing with the argument.
#
# Arguments:
#  msg		the message to log.
#
# Results:
#  None.

proc procomp::nullLogProc { msg } {
}

# procomp::log --
#
#  Logging procedure: calls the current log procedure, passes $msg to it.
#
# Arguments:
#  msg		the message to log.
#
# Results:
#  None.

proc procomp::log { msg } {
    variable logProc
    $logProc $msg
}

# procomp::logError --
#
#  Error logging procedure: calls the current error log procedure, passes
#  "error: $msg" to it.
#
# Arguments:
#  msg		the error message to log.
#
# Results:
#  None.

proc procomp::logError { msg } {
    variable errorProc
    $errorProc "error: $msg"
}

# procomp::fileCompile --
#
#  Compiles a file.
#
# Arguments:
#  path		the path to the file to compile.
#  outputFile	(optional) the output file name; if not specified, determine
#		the output file name from the control structures.
#		This argument is used when fileCompile is called from
#		dirCompile.
#
# Results:
#  Returns 1 on success, 0 on failure.

proc procomp::fileCompile { path {outputFile {}} } {
    variable outPath
    variable byteCodeExtension
    variable headerType
    variable headerValue
    variable forceWrite

    set cmd compiler::compile

    # Get the preamble if any

    if {[catch {
	switch -- $headerType {
	    none {
	    }

	    auto {
		lappend cmd -preamble [getAutoHeader $path]
	    }

	    tag {
		lappend cmd -preamble [getTagHeader $path]
	    }

	    default {
		lappend cmd -preamble $headerValue
	    }
	}
    } err] == 1} {
	logError "compilation of \"$path\" failed: $err"
	return 0
    }

    set outputFile [generateOutFileName $path $outputFile]
    lappend cmd $path $outputFile

    # and finally run the compile

    if {[catch {
	if {$forceWrite == 1} {
	    file delete -force $outputFile
	}
	eval $cmd
    } err] == 1} {
	logError "compilation of \"$path\" failed: $err"
	return 0
    }

    log "compiled: $path to $outputFile"

    return 1
}

# procomp::generateOutFileName --
#
#  Generate an output file name from the input file name and the given output
#  file name. The generated path may be more complete than the one passed in.
#
# Arguments:
#  path		the path to the file to compile.
#  outputFile	(optional) the output file name; if not specified, determine
#		the output file name from the control structures.
#
# Results:
#  Returns the generated path, which may be the same as the value of
#  $outputFile

proc procomp::generateOutFileName { path {outputFile {}} } {
    variable outPath
    variable byteCodeExtension

    set tbcName [file rootname [file tail $path]]$byteCodeExtension

    if {[string compare $outputFile {}] == 0} {
	# if outPath is specified, construct the output file name from it.
	# Otherwise, the output file name is the input file name, with 
	# the .tbc extension.

	if {[info exists outPath] == 1} {
	    if {[file isdirectory $outPath] == 1} {
		set tbcName [file join $outPath $tbcName]
	    } else {
		set tbcName $outPath
	    }
	} else {
	    set tbcName [file join [file dir $path] $tbcName]
	}
    } else {
	# if the given output file is a directory, get the input file name,
	# tag the .tbc extension, and append it to the directory path.
	# If it is a file, or we can't tell, return it as is

	if {[file isdirectory $outputFile] == 1} {
	    set tbcName [file join $outputFile $tbcName]
	} else {
	    set tbcName $outputFile
	}
    }

    return $tbcName
}

# procomp::getAutoHeader --
#
#  Parses a script, extract the 'auto' header from it: all lines from the
#  beginning up to the first non-comment or the first empty line. Here, a line
#  containing only whitespace is considered to be empty.
#
# Arguments:
#  path		the path to the script to parse.
#
# Results:
#  Returns the header; may throw on error.

proc procomp::getAutoHeader { path } {
    variable tagValue

    set istm [open $path r]
    set header {}
    set continuation 0

    while {[gets $istm line] >= 0} {
	if {$continuation == 0} {
	    if {([regexp {^[ ]*#} $line] == 0) \
		    || ([regexp {^[ ]*$} $line] == 1)} {
		break
	    }
	}

	# we need to check if this line is continued on the next one

	if {[regexp {\\$} $line] == 1} {
	    set continuation 1
	} else {
	    set continuation 0
	}

	append header $line "\n"
    }
    close $istm

    append header [format "# %s\n" $tagValue]

    return $header
}

# procomp::getTagHeader --
#
#  Parses a script, extract the 'tag' header from it: all lines from the
#  beginning up to the first occurrence of the TclPro compiler include tag.
#  It is an error to parse a script that is missing the tag.
#
# Arguments:
#  path		the path to the script to parse.
#
# Results:
#  Returns the header; may throw on error.

proc procomp::getTagHeader { path } {
    variable tagPattern
    variable tagValue
    variable tagError

    set istm [open $path r]
    set header {}
    set didSeeTag 0

    while {[gets $istm line] >= 0} {
	if {[regexp $tagPattern $line] == 1} {
	    set didSeeTag 1
	    break
	}

	append header $line "\n"
    }
    close $istm

    if {$didSeeTag == 0} {
	if {$tagError == 0} {
	    return {}
	}
	error "missing header include tag"
    }

    append header [format "# %s\n" $tagValue]

    return $header
}
